File: packages/pvm-assemblyscript/assembly/host-functions/accumulate/designate.ts
import { RESULT_CODE_PANIC } from '../../config'
import {
  ACCUMULATE_ERROR_HUH,
  AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
  HostFunctionResult,
} from './base'

/**
 * DESIGNATE accumulation host function (Ω_D)
 *
 * Designates validators (only delegator can do this)
 *
 * Gray Paper Specification:
 * - Function ID: 16 (designate)
 * - Gas Cost: 10
 * - Parameters: registers[7] = o
 *   - o: validators array offset in memory
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read validators array from memory (336 bytes per validator, up to Cvalcount validators)
 * 2. Check if current service is the delegator (imX.id === imX.state.ps_delegator)
 * 3. Update staging set with new validators
 * 4. Return OK on success, HUH if not delegator, PANIC on error
 */
export class DesignateHostFunction extends BaseAccumulateHostFunction {
  functionId: u64 = u64(16) // DESIGNATE function ID
  name: string = 'designate'
  gasCost: u64 = u64(10)

  // Gray Paper constants
  VALIDATOR_SIZE: i32 = 336 // bytes per validator

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const registers = context.registers
    const ram = context.ram
    const implications = context.implications
    const numValidators = context.numValidators // Get from config via context
    
    // Get the current implications context
    const imX = implications.regular

    // Extract parameters from registers
    const validatorsOffset = u64(registers[7])

    // Read validators array from memory (336 bytes per validator, up to Cvalcount validators)
    // Gray Paper: sequence[Cvalcount]{valkey} where Cvalcount comes from config
    // Gray Paper: v = sequence{memory[o+336i:336] for i in valindex} when readable, error otherwise
    const totalSize = this.VALIDATOR_SIZE * i32(numValidators)

    const readResult_validators = ram.readOctets(
      u32(validatorsOffset),
      u32(totalSize),
    )
    
    // Gray Paper: (panic, registers_7, stagingset) when v = error
    // Check memory read FIRST (before delegator check) per Gray Paper order
    // DO NOT modify registers[7] - it must remain unchanged on panic
    if (readResult_validators.faultAddress !== 0 || readResult_validators.data === null) {
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    const validatorsData = readResult_validators.data!

    // Gray Paper: (continue, HUH, stagingset) otherwhen imX_id ≠ (imX_state)_ps_delegator
    // Check delegator AFTER successful memory read per Gray Paper order
    if (imX.id !== u64(imX.state.delegator)) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_HUH)
      return new HostFunctionResult(255) // continue execution
    }

    // Parse validators array (336 bytes per validator)
    // Note: For AssemblyScript, we store raw validator data as Uint8Array
    // Full implementation would decode ValidatorPublicKeys structure
    const validators: Uint8Array[] = []
    for (let i: i32 = 0; i < i32(numValidators); i++) {
      const validatorData = validatorsData.slice(
        i * this.VALIDATOR_SIZE,
        (i + 1) * this.VALIDATOR_SIZE,
      )
      // Note: Would decode ValidatorPublicKeys here if codec available
      // For now, store raw bytes
      validators.push(validatorData)
    }

    // Update staging set with new validators
    // Gray Paper: (imX'.state).ps_stagingset = v
    imX.state.stagingset = validators

    // Set success result
    this.setAccumulateSuccess(registers)
    return new HostFunctionResult(255) // continue execution
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/designate.rs
//! DESIGNATE accumulation host function (Ω_D). Gray Paper: function ID 16.
//! r7 = validators array offset (o). 1:1 with AS designate.ts.

use crate::config::FUNC_DESIGNATE;
use crate::host_functions::accumulate::base::{self, codes, DEFAULT_NUM_VALIDATORS, VALIDATOR_SIZE};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

pub struct DesignateHostFunction;

impl HostFunction for DesignateHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_DESIGNATE
    }
    fn name(&self) -> &'static str {
        "designate"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let validators_offset = context.registers[7];
        let num_validators = context
            .num_validators
            .unwrap_or(DEFAULT_NUM_VALIDATORS);
        let total_size = (VALIDATOR_SIZE * num_validators) as u32;

        let read_result = context.ram.read_octets(validators_offset as u32, total_size);
        if read_result.fault_address != 0 || read_result.data.is_none() {
            crate::host_log!(
                "[hostfn] designate PANIC: validators read fault (offset={}, len={}, fault_address={})",
                validators_offset, total_size, read_result.fault_address
            );
            return HostFunctionResult::panic();
        }
        let validators_data = read_result.data.unwrap();
        if validators_data.len() != total_size as usize {
            crate::host_log!(
                "[hostfn] designate PANIC: validators length mismatch (got {}, expected {})",
                validators_data.len(),
                total_size as usize
            );
            return HostFunctionResult::panic();
        }

        // AS: imX.id !== imX.state.delegator → HUH. Single imX.state (accumulation_state).
        let state = context
            .accumulation_state
            .as_deref_mut()
            .expect("designate: accumulation context must have accumulation_state");
        let current_service_id = context
            .service_id
            .expect("designate: accumulation context must have service_id");
        if current_service_id != u64::from(state.delegator) {
            base::set_accumulate_error(context.registers, codes::HUH);
            return HostFunctionResult::continue_execution();
        }

        let validator_size = VALIDATOR_SIZE as usize;
        let mut validators = Vec::with_capacity(num_validators as usize);
        for i in 0..(num_validators as usize) {
            let start = i * validator_size;
            let end = start + validator_size;
            validators.push(validators_data[start..end].to_vec());
        }
        state.stagingset = validators;

        base::set_accumulate_success(context.registers, codes::OK);
        HostFunctionResult::continue_execution()
    }
}
